File: Backend_Java/src/main/java/com/bookstore/entity/Coupon.java
package com.bookstore.entity;

import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "coupons")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 50)
    private String code;

    @Column(length = 255)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private CouponType type;

    @Column(name = "discount_value", precision = 12, scale = 2, nullable = false)
    private BigDecimal discountValue;

    @Column(name = "min_order_amount", precision = 12, scale = 2)
    @Builder.Default
    private BigDecimal minOrderAmount = BigDecimal.ZERO;

    @Column(name = "max_discount", precision = 12, scale = 2)
    @Builder.Default
    private BigDecimal maxDiscount = BigDecimal.ZERO;

    @Column(name = "start_date")
    private LocalDateTime startDate;

    @Column(name = "end_date")
    private LocalDateTime endDate;

    @Column(name = "usage_limit")
    @Builder.Default
    private Integer usageLimit = 0;

    @Column(name = "used_count")
    @Builder.Default
    private Integer usedCount = 0;

    @Column(name = "per_user_limit")
    @Builder.Default
    private Integer perUserLimit = 1;

    @Column(name = "is_active")
    @Builder.Default
    private Boolean isActive = true;

    @Column(name = "applicable_categories", columnDefinition = "TEXT")
    private String applicableCategories;

    @Column(name = "applicable_products", columnDefinition = "TEXT")
    private String applicableProducts;

    @Column(name = "is_public")
    @Builder.Default
    private Boolean isPublic = true;

    @Column(name = "created_by")
    private Long createdBy;

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public boolean isValid() {
        if (!isActive) return false;
        LocalDateTime now = LocalDateTime.now();
        if (startDate != null && now.isBefore(startDate)) return false;
        if (endDate != null && now.isAfter(endDate)) return false;
        if (usageLimit > 0 && usedCount >= usageLimit) return false;
        return true;
    }

    public boolean isExpired() {
        return endDate != null && LocalDateTime.now().isAfter(endDate);
    }
}

File: Frontend_NextJS/src/lib/coupon.ts
import { api } from "./api";

export interface Coupon {
  id: number;
  code: string;
  description: string;
  type: "PERCENTAGE" | "FIXED_AMOUNT" | "FREE_SHIPPING";
  discountValue: number;
  minOrderAmount: number;
  maxDiscount: number;
  startDate: string;
  endDate: string;
  usageLimit: number;
  usedCount: number;
  perUserLimit: number;
  isActive: boolean;
  isPublic: boolean;
  isValid: boolean;
  isExpired: boolean;
  discountDisplay: string;
  createdAt: string;
}

export interface CouponRequest {
  code: string;
  description: string;
  type: "PERCENTAGE" | "FIXED_AMOUNT" | "FREE_SHIPPING";
  discountValue: number;
  minOrderAmount?: number;
  maxDiscount?: number;
  startDate: string;
  endDate: string;
  usageLimit?: number;
  perUserLimit?: number;
  isPublic?: boolean;
  applicableCategories?: number[];
  applicableProducts?: number[];
}

export interface CouponValidationRequest {
  code: string;
  orderTotal?: number;
}

export const couponApi = {
  getAvailableCoupons: async () => {
    const response = await api.get<{ data: Coupon[] }>("/coupons/available");
    return response.data.data;
  },

  getAllCoupons: async (page = 0, size = 20) => {
    const response = await api.get<{ data: Coupon[] }>(
      `/coupons?page=${page}&size=${size}`
    );
    return response.data.data;
  },

  getCouponById: async (id: number) => {
    const response = await api.get<{ data: Coupon }>(`/coupons/${id}`);
    return response.data.data;
  },

  createCoupon: async (request: CouponRequest) => {
    const response = await api.post<{ data: Coupon }>("/coupons", request);
    return response.data.data;
  },

  updateCoupon: async (id: number, request: CouponRequest) => {
    const response = await api.put<{ data: Coupon }>(`/coupons/${id}`, request);
    return response.data.data;
  },

  deleteCoupon: async (id: number) => {
    await api.delete(`/coupons/${id}`);
  },

  validateCoupon: async (code: string, orderTotal?: number) => {
    const response = await api.post<{ data: Coupon }>("/coupons/validate", {
      code,
      orderTotal,
    });
    return response.data.data;
  },

  searchCoupons: async (keyword: string, page = 0, size = 20) => {
    const response = await api.get<{ data: Coupon[] }>(
      `/coupons/search?keyword=${keyword}&page=${page}&size=${size}`
    );
    return response.data.data;
  },
};
